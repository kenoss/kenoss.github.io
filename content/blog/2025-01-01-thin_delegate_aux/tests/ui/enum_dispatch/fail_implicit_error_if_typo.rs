#[enum_dispatch::enum_dispatch]
trait ShapeI {
    fn area(&self) -> usize;
}

struct Rect {
    width: usize,
    height: usize,
}

// enum_dispatch can't detect typos.
#[enum_dispatch::enum_dispatch(ShapeITypo)]
enum Shape {
    Rect(Rect),
}

impl ShapeI for Rect {
    fn area(&self) -> usize {
        self.width * self.height
    }
}

fn main() {
    let rect = Rect { width: 2, height: 3 };
    assert_eq!(rect.area(), 6);
    let shape = Shape::Rect(rect);
    assert_eq!(shape.area(), 6);
}
