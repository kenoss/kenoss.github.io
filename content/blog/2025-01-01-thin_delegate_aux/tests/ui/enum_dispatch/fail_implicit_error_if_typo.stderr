error[E0599]: no method named `area` found for enum `Shape` in the current scope
  --> tests/ui/enum_dispatch/fail_implicit_error_if_typo.rs:27:22
   |
13 | enum Shape {
   | ---------- method `area` not found for this enum
...
27 |     assert_eq!(shape.area(), 6.0);
   |                      ^^^^ method not found in `Shape`
   |
note: the method `area` exists on the type `Rect`
  --> tests/ui/enum_dispatch/fail_implicit_error_if_typo.rs:3:5
   |
3  |     fn area(&self) -> f64;
   |     ^^^^^^^^^^^^^^^^^^^^^^
   = help: items from traits can only be used if the trait is implemented and in scope
note: `ShapeI` defines an item `area`, perhaps you need to implement it
  --> tests/ui/enum_dispatch/fail_implicit_error_if_typo.rs:2:1
   |
2  | trait ShapeI {
   | ^^^^^^^^^^^^
